src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateGroup.cs

```diff
@@ -0,0 +1,283 @@
+using Microsoft.TemplateEngine.Cli.HelpAndUsage;
+using Microsoft.TemplateEngine.Edge.Template;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.TemplateEngine.Cli.TemplateResolution
+{
+    /// <summary>
+    /// The class represents template group. Templates in single group:<br/>
+    /// - should same group identity
+    /// - should have different template identity <br/>
+    /// - same short name (however different short names are also supported) <br/>
+    /// - the templates may have different languages and types <br/>
+    /// - the templates should have different precedence value in case same language is used <br/>
+    /// - the templates in the group may have different parameters and different choices for parameter symbols defined<br/>
+    /// In case the template does not have group identity defined it represents separate template group with single template.
+    /// </summary>
+    internal class TemplateGroup
+    {
+        internal TemplateGroup(string groupIdentity, IEnumerable<ITemplateMatchInfo> templates)
+        {
+            GroupIdentity = groupIdentity;
```

Comment:
should `groupIdentity` be validated?

Reply:
Potentially it could be good improvement to validate if templates passed have the same groupIdentity defined, or possibly even remove group identity as the parameter.